### DISCUSSION TITLE
Use InputEventKey instead of InputEvent in _unhandled_key_input function

### DISCUSSION THREAD

**Discussion by ghost** [godotengine/godot-proposals#6968] [Engine Core]
### Describe the project you are working on

See the title

### Describe the problem or limitation you are having in your project

See the title

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

See the title

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

See the title

### If this enhancement will not be used often, can it be worked around with a few lines of script?

See the title

### Is there a reason why this should be core and not an add-on in the asset library?

See the title

**Comment by Calinou** [comment]
Out of curiosity, why would you need this?

PS: The fields in the proposal template are not optional. You need to fill *all* of them.

**Reply by ghost** [reply]
Godot has `_unhandled_input` function that also gets `InputEvent` and name of` _unhadled_key_input` function says that it can't get something other than `InputEventKey`

**Reply by AThousandShips** [reply]
While it might technically be correct I don't think being that specific is worth doing a compatibility breaking changes

**Reply by Zireael07** [reply]
... btw it's too late for compat breaking changes anyway, it'd have to wait for 5.0 which would be years away

**Reply by AThousandShips** [reply]
There's some room for compat breakage between minor versions with the new faster release schedule, this one would be a relatively serious one, affecting both script and extensions and modules

But the fact remains that this change isn't supported, see the comment below

**Comment by KoBeWi** [comment]
The reason why `_unhadled_key_input` takes InputEvent is that the event can also be InputEventShortcut. It was InputEventKey before Command Palette was implemented.

**Reply by MajorMcDoom** [reply]
This really should be mentioned in the docs. I had to come here to find out. Up until now, I'd just directly cast into `InputEventKey` instead of checking the type first, which is an unsafe operation in light of this new info.

The docs say: `Called when an InputEventKey hasn't been consumed by _input() or any GUI Control item.`

In fact, it explicitly redirects you away from this method if you want shortcuts: `For shortcuts, consider using _shortcut_input() instead.`

**Reply by ghost** [reply]
Yeah, fully agree with @MajorMcDoom. [Current docs](https://docs.godotengine.org/en/4.2/classes/class_node.html#class-node-private-method-unhandled-key-input) of `_unhadled_key_input` mentions only `InputEventKey`

**Reply by KoBeWi** [reply]
See https://github.com/godotengine/godot/issues/83581
Apparently the method was changed and no longer receives InputEventShortcut, but the argument is kept for compatibility reasons.
